src/ast/passes/semantic_analyser.cpp

```diff
@@ -216,6 +216,303 @@ class SemanticAnalyser : public Visitor<SemanticAnalyser> {
 
 } // namespace
 
+struct assign_checks {
+  bool want_map = false;
+  bool want_var = false;
+  bool want_map_key = false;
+};
+
+// clang-format off
+static const std::map<std::string, assign_checks> CALL_ASSIGN_SPEC = {
+  { "avg", { .want_map = true } },
+  { "cat", {} },
+  { "clear", {} },
+  { "count", { .want_map = true } },
+  { "debugf", {} },
+  { "delete", {} },
+  { "exit", {} },
+  { "hist", { .want_map = true } },
+  { "join", {} },
+  { "lhist", { .want_map = true } },
+  { "max", { .want_map = true } },
+  { "min", { .want_map = true } },
+  { "override", {} },
+  { "print", {} },
+  { "printf", {} },
+  { "signal", {} },
+  { "skboutput", {.want_var = true} },
+  { "stats", { .want_map = true } },
+  { "sum", { .want_map = true } },
+  { "system", {} },
+  { "time", {} },
+  { "zero", {} },
+};
+
+struct arg_type_spec {
+  Type type = Type::integer;
+  bool literal = false;
+  // This means that this check is only valid if there are
+  // this exact number of args in the call
+  size_t nargs = 0;
+  // This indicates that this is just a placeholder
+  // as we use the index in the vector of arg_type_spec
+  // as the number argument to check
+  bool no_check = false;
```

Comment:
Why not use the arg index here directly? IIUC, it's not used for anything else than the check itself and it would allow us to only specify and iterate indices which need checking.

Reply:
@viktormalik Can you explain a bit more. I don't quite understand what you're saying. Also if you look at the spec, you'll see where `no_check` is used (which is only a few places). I had originally tried to define the types field as `std::vector<std::optional<arg_type_spec>> types` to avoid this field but then I would have to wrap all non `nullopt` values in `make_optional`, which was gross. I also originally had a `num` field in this struct to indicate the arg index but thought it was also kind of extra cruft when I could just use the index of into the array.

Reply:
Yeah, I meant having an explicit `num` field in the struct.

It's just that to me, it'd easier to "parse" that large initializer in my head if every argument had its index explicitly specified. But I understand that using the vector index is less code and that we don't have that many places where `no_check` is used so I'm not going to push this as it may be just my perception.

Also, if we keep it, maybe `skip_check` would be a bit better?

Reply:
I like `skip_check`.